File: sdk/php/src/HttpClient.php
<?php

namespace HiTechCloud\Tools;

use GuzzleHttp\Client;

class HttpClient
{
    private $client;
    private $maxRetries;

    public function __construct(string $baseUrl = 'https://api-tools.hitechcloud.vn', ?string $apiKey = null, int $timeout = 30, int $maxRetries = 3)
    {
        $this->maxRetries = $maxRetries;
        $headers = ['Accept' => 'application/json', 'User-Agent' => 'HiTechCloud-Tools-PHP-SDK/1.0.0'];
        if ($apiKey) $headers['X-API-Key'] = $apiKey;
        $this->client = new Client(['base_uri' => rtrim($baseUrl, '/'), 'timeout' => $timeout, 'headers' => $headers]);
    }

    public function setApiKey(string $key): void { $this->client = new Client(array_merge($this->client->getConfig(), ['headers' => array_merge($this->client->getConfig()['headers'] ?? [], ['X-API-Key' => $key])])); }

    public function get(string $path, array $params = []): array { return $this->request('GET', $path, ['query' => $params]); }
    public function post(string $path, array $data = []): array { return $this->request('POST', $path, ['json' => $data]); }

    private function request(string $method, string $path, array $options = []): array
    {
        for ($attempt = 0; $attempt < $this->maxRetries; $attempt++) {
            try {
                $response = $this->client->request($method, $path, $options);
                return json_decode($response->getBody()->getContents(), true);
            } catch (\GuzzleHttp\Exception\RequestException $e) {
                if ($e->hasResponse()) {
                    $status = $e->getResponse()->getStatusCode();
                    if ($status === 429 && $attempt < $this->maxRetries - 1) { sleep((int)$e->getResponse()->getHeaderLine('Retry-After') ?: 60); continue; }
                    if ($status >= 400) {
                        $body = json_decode($e->getResponse()->getBody()->getContents(), true);
                        $msg = $body['error'] ?? $body['message'] ?? "HTTP $status";
                        throw new \RuntimeException(is_array($msg) ? implode('; ', $msg) : $msg, $status);
                    }
                }
                if ($attempt < $this->maxRetries - 1) { sleep(pow(2, $attempt)); continue; }
                throw $e;
            }
        }
        throw new \RuntimeException('Request failed after retries');
    }
}

File: sdk/dart/lib/src/http_client.dart
import 'dart:convert';
import 'package:http/http.dart' as http;

class HttpClient {
  final String baseUrl;
  final String apiKey;
  final int timeout;
  final int maxRetries;

  HttpClient({required this.baseUrl, required this.apiKey, this.timeout = 30, this.maxRetries = 3});

  

  Future<Map<String, dynamic>> get(String path, [Map<String, dynamic>? params]) async {
    final uri = _buildUri(path, params);
    final response = await http.get(uri, headers: _headers());
    return jsonDecode(response.body) as Map<String, dynamic>;
  }

  Future<Map<String, dynamic>> post(String path, [Map<String, dynamic>? body]) async {
    final uri = _buildUri(path);
    final response = await http.post(uri, headers: _headers(), body: jsonEncode(body ?? {}));
    return jsonDecode(response.body) as Map<String, dynamic>;
  }

  Future<Map<String, dynamic>> put(String path, [Map<String, dynamic>? body]) async {
    final uri = _buildUri(path);
    final response = await http.put(uri, headers: _headers(), body: jsonEncode(body ?? {}));
    return jsonDecode(response.body) as Map<String, dynamic>;
  }

  Future<Map<String, dynamic>> delete(String path) async {
    final uri = _buildUri(path);
    final response = await http.delete(uri, headers: _headers());
    return jsonDecode(response.body) as Map<String, dynamic>;
  }

  Uri _buildUri(String path, [Map<String, dynamic>? params]) {
    var uri = Uri.parse('$baseUrl$path');
    if (params != null) {
      final filtered = Map<String, dynamic>.from(params)..removeWhere((_, v) => v == null);
      if (filtered.isNotEmpty) uri = uri.replace(queryParameters: filtered.map((k, v) => MapEntry(k, '$v')));
    }
    return uri;
  }

  Map<String, String> _headers() {
    final h = <String, String>{'Content-Type': 'application/json', 'Accept': 'application/json'};
    h['X-API-Key'] = apiKey;
    return h;
  }
}
